File: cplusplus/src/crypto/diffie_hellman.cpp
#include "crypto.hpp"
#include "sodium.h"

namespace autograph {

bool diffie_hellman(unsigned char *shared_secret,
                    const unsigned char *our_private_key,
                    const unsigned char *their_public_key) {
  return crypto_scalarmult(shared_secret, our_private_key, their_public_key) ==
         0;
}

}  // namespace autograph

File: typescript/src/crypto/diffie-hellman.ts
import { diffieHellman as deriveSharedSecret } from 'stedy'
import { importPrivateKey, importPublicKey } from '../utils'

const diffieHellman = async (
  ourPrivateKey: BufferSource,
  theirPublicKey: BufferSource
) =>
  deriveSharedSecret(
    await importPrivateKey(ourPrivateKey),
    await importPublicKey(theirPublicKey),
    32
  )

export default diffieHellman
